### PR TITLE
[IMP] core: allow skipping field recomputitation on installation

### PR DESCRIPTION
**Pull request by yelizariev**
When a module adds new computated field, it may take significant time to make
initial computation. Before this commit, the only way to avoid this was using
`_auto_init`. Example:

https://github.com/odoo/odoo/blob/ca0d8978b6585fe4f96100df56b8a7346468dc6c/addons/l10n_latam_invoice_document/models/account_move.py#L42-L44

https://github.com/odoo/enterprise/blob/46812d9abb3eb6e2b5ac565b4817548317c4a354/l10n_mx_edi_extended/models/account_move.py#L79-L82

But such method requires creating columns manually.

This commit allows to use `_init_column` method instead by adding return value
``skip_recomputation``

As a bonus, this allows make some cleanup in odoo/fields.py and move code
"optimization for computing simple related fields like 'foo_id.bar'" to _init_column

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -2555,6 +2579,21 @@ def _init_column(self, column_name):
...
+            SET "{model_field}" = y."{comodel_field}"
+            FROM "{comodel_table}" AS y
+            WHERE x."{join_field}" = y.id
+        """.format(
+            model_table=self._table,
+            model_field=field.name,
+            comodel_table=comodel._table,
+            comodel_field=field.related[1],
+            join_field=field.related[0],
+        ))
```
Comment (yelizariev) [inline odoo/models.py:2595]: runbot is read becase of this:

> Possible SQL injection risk. (E8501) at odoo/odoo/models.py:2584

Comment (pedrobaeza) [reply odoo/models.py:2595]: I use `psycopg.sql.SQL` for bulding such queries. Do you know how to use it?

Comment (pedrobaeza) [reply odoo/models.py:2595]: Quick translation (maybe not full accurate):

```python
        from psycopg2 import sql
        self.env.cr.execute(sql.SQL("""
            UPDATE {model_table} AS x
            SET {model_field} = y.{comodel_field}
            FROM {comodel_table} AS y
            WHERE x.{join_field} = y.id
        """).format(
            model_table=sql.Identifier(self._table),
            model_field=sql.Identifier(field.name),
            comodel_table=sql.Identifier(comodel._table),
            comodel_field=sql.Identifier(field.related[1]),
            join_field=sql.Identifier(field.related[0]),
        ))
```

Comment (yelizariev) [reply odoo/models.py]: I did as you suggested and it works :+1:

**Comment by yelizariev** [comment]
@Elkasitu

**Comment by yelizariev** [comment]
Updated according to changes in https://github.com/odoo/odoo/commit/1abe965b59a6acc24b93c25c07c2f9151af4924a

File: odoo/fields.py [hunk 2]
```diff
@@ -884,13 +884,9 @@ def update_db(self, model, columns):
         # create/update the column, not null constraint; the index will be
         # managed by registry.check_indexes()
         self.update_db_column(model, column)
-        skip_recomputation = self.update_db_notnull(model, column)
+        initialized = self.update_db_notnull(model, column)
```
Comment (yelizariev) [inline odoo/fields.py]: :+1: 
I felt that ``skip_recomputation`` is something like a variable with negation

File: odoo/addons/test_inherit/tests/test_inherit.py [hunk 3]
```diff
@@ -174,3 +174,12 @@ def test_xml_ids(self):
...
 
+
+class TestInitColumn(common.TransactionCase):
+    def test_init_column(self):
+        # the installation of this module should:
+        #  - use the compute method to initialize field 'test_init_computed'
+        #  - bypass the compute method to initialize field 'test_init_optimized'
+        partner = self.env.user.partner_id
+        self.assertEqual(partner.test_init_computed, 'computed')
+        self.assertEqual(partner.test_init_optimized, 'optimized')
```
Comment (yelizariev) [inline odoo/addons/test_inherit/tests/test_inherit.py:185]: This seems not related to inherit and if someone wants to find test related to field initialization they will not check ``test_inherit`` module. Maybe use ``test_new_api`` instead?

File: odoo/fields.py [hunk 4]
```diff
@@ -953,22 +938,57 @@ def add_not_null():
         elif not self.required and has_notnull:
             sql.drop_not_null(model._cr, model._table, self.name)
 
-    def update_db_related(self, model):
+        return initialized
+
+    def initialize(self, model):
+        """ Possibly initialize the field, and return ``True`` if it has been
+        actually done.
+        """
```
Comment (yelizariev) [inline odoo/fields.py:946]: :+1:

File: odoo/addons/test_inherit/tests/test_inherit.py (hunk 3 above)
Comment (rco-odoo) [reply odoo/addons/test_inherit/tests/test_inherit.py:185]: It is related to inheritance in the way that inheritance adds columns to a non-empty table.
There was already an extension of model `res.partner` in that module, so this made the diff minimal for the test...

**Comment by yelizariev** [comment]
runbot is red because of this: https://github.com/odoo/odoo/pull/71338

**Comment by yelizariev** [comment]
@rco-odoo new version changes something with recomputed fields

in `hr_holidays` module there are storable computed readonly=False fields:

https://github.com/odoo/odoo/blob/eca6e539a6f62303b4fb1bd5406082d5df5d589f/addons/hr_holidays/models/hr_leave.py#L150-L155

After this PR, recomputing those fields triggers constrains which is not supposed checking changes made by computed method.

https://github.com/odoo/odoo/blob/eca6e539a6f62303b4fb1bd5406082d5df5d589f/addons/hr_holidays/models/hr_leave.py#L586-L592

<details>

<summary>stack</summary>

```
Failed to initialize database `7957527-master-all`.
Traceback (most recent call last):
  File "/data/build/odoo/odoo/api.py", line 867, in get
    return field_cache[record._ids[0]]
KeyError: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/fields.py", line 1068, in __get__
    value = env.cache.get(record, self)
  File "/data/build/odoo/odoo/api.py", line 870, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'hr.leave(1,).validation_type'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/service/server.py", line 1246, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "/data/build/odoo/odoo/modules/registry.py", line 88, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/data/build/odoo/odoo/modules/loading.py", line 471, in load_modules
    loaded_modules, update_module, models_to_check)
  File "/data/build/odoo/odoo/modules/loading.py", line 364, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/data/build/odoo/odoo/modules/loading.py", line 198, in load_module_graph
    registry.init_models(cr, model_names, {'module': package.name}, new_install)
  File "/data/build/odoo/odoo/modules/registry.py", line 437, in init_models
    env['base'].flush()
  File "/data/build/odoo/odoo/models.py", line 5503, in flush
    self.recompute()
  File "/data/build/odoo/odoo/models.py", line 5976, in recompute
    process(field)
  File "/data/build/odoo/odoo/models.py", line 5960, in process
    field.recompute(recs)
  File "/data/build/odoo/odoo/fields.py", line 1252, in recompute
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1274, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 409, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4129, in _compute_field_value
    getattr(self, field.compute)()
  File "/data/build/odoo/addons/hr_holidays/models/hr_leave.py", line 312, in _compute_state
    holiday.state = 'confirm' if holiday.validation_type != 'no_validation' else 'draft'
  File "/data/build/odoo/odoo/fields.py", line 1117, in __get__
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1274, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 409, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4131, in _compute_field_value
    field.compute(self)
  File "/data/build/odoo/odoo/fields.py", line 606, in _compute_related
    values = [first(value[name]) for value in values]
  File "/data/build/odoo/odoo/fields.py", line 606, in <listcomp>
    values = [first(value[name]) for value in values]
  File "/data/build/odoo/odoo/models.py", line 5747, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/data/build/odoo/odoo/fields.py", line 2598, in __get__
    return super().__get__(records, owner)
  File "/data/build/odoo/odoo/fields.py", line 1065, in __get__
    self.recompute(record)
  File "/data/build/odoo/odoo/fields.py", line 1252, in recompute
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1274, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 409, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4129, in _compute_field_value
    getattr(self, field.compute)()
  File "/data/build/odoo/addons/hr_holidays/models/hr_leave.py", line 433, in _compute_from_employee_id
    holiday.manager_id = holiday.employee_id.parent_id.id
  File "/data/build/odoo/odoo/fields.py", line 2598, in __get__
    return super().__get__(records, owner)
  File "/data/build/odoo/odoo/fields.py", line 1065, in __get__
    self.recompute(record)
  File "/data/build/odoo/odoo/fields.py", line 1252, in recompute
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1274, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 409, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4136, in _compute_field_value
    self.filtered('id')._validate_fields(fnames)
  File "/data/build/odoo/odoo/models.py", line 1357, in _validate_fields
    check(self)
  File "/data/build/odoo/addons/hr_holidays/models/hr_leave.py", line 565, in _check_date
    ('date_from', '<', holiday.date_to),
  File "/data/build/odoo/odoo/fields.py", line 1065, in __get__
    self.recompute(record)
  File "/data/build/odoo/odoo/fields.py", line 1252, in recompute
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1274, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 409, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4136, in _compute_field_value
    self.filtered('id')._validate_fields(fnames)
  File "/data/build/odoo/odoo/models.py", line 1357, in _validate_fields
    check(self)
  File "/data/build/odoo/addons/hr_holidays/models/hr_leave.py", line 592, in _check_date_state
    raise ValidationError(_("This modification is not allowed in the current state."))
odoo.exceptions.ValidationError: This modification is not allowed in the current state.
```


</details>

Any hint what to do?
